File: src/second_iteration/map_solution/RockPaperScissorsSolver.java
package second_iteration.map_solution;

import java.util.HashMap;
import java.util.Map;

import static second_iteration.map_solution.Gesture.*;

public class RockPaperScissorsSolver {

	private static final Map<Play, String> plays = new HashMap<>();
	static {
		plays.put(play(ROCK, ROCK), "Tie.");
		plays.put(play(ROCK, PAPER), "Rock loses against paper.");
		plays.put(play(ROCK, SCISSORS), "Rock wins against scissors.");
		plays.put(play(PAPER, ROCK), "Paper wins against rock.");
		plays.put(play(PAPER, PAPER), "Tie.");
		plays.put(play(PAPER, SCISSORS), "Paper loses against scissors.");
		plays.put(play(SCISSORS, ROCK), "Scissors lose against rock.");
		plays.put(play(SCISSORS, PAPER), "Scissors win against paper.");
		plays.put(play(SCISSORS, SCISSORS), "Tie.");
	}

	public static String solve(Gesture leftGesture, Gesture rightGesture) {
		return plays.get(play(leftGesture, rightGesture));
	}

	private static Play play(Gesture left, Gesture right) {
		return new Play(left, right);
	}

	private static class Play {
		private final Gesture left;
		private final Gesture right;

		public Play(Gesture left, Gesture right) {
			this.left = left;
			this.right = right;
		}

		@Override
		public boolean equals(Object obj) {
			return obj instanceof Play && left == ((Play) obj).left && right == ((Play) obj).right;
		}

		@Override
		public int hashCode() {
			return left.hashCode() * right.hashCode();
		}
	}
}

File: php/src/second_iteration/map_solution/RockPaperScissorsSolver.php
<?php

namespace second_iteration\map_solution;

class RockPaperScissorsSolver
{

    private static $plays = null;

    /**
     * @return string[]
     */
    public static function plays()
    {
        if (self::$plays == null) {
            self::$plays = array(
                Gesture::ROCK . Gesture::ROCK => "Tie.",
                Gesture::ROCK . Gesture::PAPER => "Rock loses against paper.",
                Gesture::ROCK . Gesture::SCISSORS => "Rock wins against scissors.",
                Gesture::PAPER . Gesture::ROCK => "Paper wins against rock.",
                Gesture::PAPER . Gesture::PAPER => "Tie.",
                Gesture::PAPER . Gesture::SCISSORS => "Paper loses against scissors.",
                Gesture::SCISSORS . Gesture::ROCK => "Scissors lose against rock.",
                Gesture::SCISSORS . Gesture::PAPER => "Scissors win against paper.",
                Gesture::SCISSORS . Gesture::SCISSORS => "Tie.",
            );
        }
        return self::$plays;
    }

    /**
     * @param Play $play
     * @return string
     */
    public static function getPlay(Play $play)
    {
        return self::plays()[(string)$play];
    }

    /**
     * @param string $leftGesture
     * @param string $rightGesture
     * @return string
     */
    public static function solve($leftGesture, $rightGesture)
    {
        return self::getPlay(self::play($leftGesture, $rightGesture));
    }

    /**
     * @param string $left
     * @param string $right
     * @return Play
     */
    private static function play($left, $right)
    {
        return new Play($left, $right);
    }
}
